source/game_sa/CFont.cpp

```diff
@@ -56,206 +90,582 @@ void CFont::Initialise() {
-}
-
+    // info about tokens: https://gtamods.com/wiki/GXT#Tokens
+
+    char* next = ++text;
+
+    auto ApplyStyle = [&](eHudColours hudColor) {
+        if (!isBlip)
+            color = HudColour.GetRGBA(hudColor, color.a);
+
+        if (tag)
+            *tag = *next;
+    };
+
+    switch (*next) {
+    case '<':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_LEFT;
+        break;
+    case '>':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_RIGHT;
+        break;
+    case 'A':
+    case 'a':
+        m_nExtraFontSymbolId = EXSYMBOL_L3;
+        break;
+    case 'B':
+    case 'b':
+        ApplyStyle(HUD_COLOUR_DARK_BLUE);
+        break;
+    case 'C':
+    case 'c':
+        m_nExtraFontSymbolId = EXSYMBOL_R3;
+        break;
+    case 'D':
+    case 'd':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_DOWN;
+        break;
+    case 'G':
+    case 'g':
+        ApplyStyle(HUD_COLOUR_GREEN);
+        break;
+    case 'H':
+    case 'h':
+        if (!isBlip) {
+            color = {
+                    std::min<uint8>(color.r * 1.5f, 255.0f),
+                    std::min<uint8>(color.g * 1.5f, 255.0f),
+                    std::min<uint8>(color.b * 1.5f, 255.0f),
+                    color.a
+            };
+        }
+
+        if (tag)
+            *tag = *next;
+        break;
+    case 'J':
+    case 'j':
+        m_nExtraFontSymbolId = EXSYMBOL_R1;
+        break;
+    case 'K':
+    case 'k':
+        m_nExtraFontSymbolId = EXSYMBOL_KEY;
+        break;
+    case 'M':
+    case 'm':
+        m_nExtraFontSymbolId = EXSYMBOL_L2;
+        break;
+    case 'N':
+    case 'n':
+        m_bNewLine = true;
+        break;
+    case 'O':
+    case 'o':
+        m_nExtraFontSymbolId = EXSYMBOL_CIRCLE;
+        break;
+    case 'P':
+    case 'p':
+        ApplyStyle(HUD_COLOUR_PURPLE);
+        break;
+    case 'Q':
+    case 'q':
+        m_nExtraFontSymbolId = EXSYMBOL_SQUARE;
+        break;
+    case 'R':
+    case 'r':
+        ApplyStyle(HUD_COLOUR_RED);
+        break;
+    case 'S':
+    case 's':
+        ApplyStyle(HUD_COLOUR_LIGHT_GRAY);
+        break;
+    case 'T':
+    case 't':
+        m_nExtraFontSymbolId = EXSYMBOL_TRIANGLE;
+        break;
+    case 'U':
+    case 'u':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_UP;
+        break;
+    case 'V':
+    case 'v':
+        m_nExtraFontSymbolId = EXSYMBOL_R2;
+        break;
+    case 'W':
+    case 'w':
+        ApplyStyle(HUD_COLOUR_LIGHT_GRAY);
+        break;
+    case 'X':
+    case 'x':
+        m_nExtraFontSymbolId = EXSYMBOL_CROSS;
+        break;
+    case 'Y':
+    case 'y':
+        ApplyStyle(HUD_COLOUR_CREAM);
+        break;
+    case 'l':
+        ApplyStyle(HUD_COLOUR_BLACK);
+        break;
+    default:
+        break;
+    }
+
+    if (*next != '~') {
+        // skip text to the next '~' character.
+        char x;
+        do
+            x = (next++)[1];
+        while (x != '~');
```

Comment:
We can do this the C way:
```suggestion
        next = strchr(next, '~');
```
Or (With this we can remove the `if`)
Also, this is most likely what R* used (at least from my experience):
```suggestion
        for(; *next && *next != '~'; next++);
```